add logs and info read endpoints

deitch opened the pull request:
Endpoints to read all of the logs and info.

This does not allow watching... yet. Future PR will.

deitch commented:
Update: now allows watching, but still requires docs for it and a little more testing. Do not merge until I signal here that it is ready.

deitch commented:
And docs added. This is ready to go in.

rvs commented:
This is super awesome -- let me review @deitch

deitch commented:
Hold on a moment; adding the CLI version of getting...

deitch commented:
OK @rvs added docs and cli commands as well. Go for it.

rvs commented on pkg/server/adminHandler.go:
```diff
@@ -4,22 +4,34 @@
+
+const (
+	StreamHeader = "X-Stream"
+	StreamValue  = "true"
```
LOL -- I love how this constant reads "extreme"! ;-)

deitch replied:
Hohoho! I didn't think of that!

rvs commented on cmd/info.go:
```diff
@@ -0,0 +1,46 @@
+		if err != nil {
+			log.Fatalf("error reading URL %s: %v", u, err)
+		}
+		if _, err := io.Copy(os.Stdout, response.Body); err != nil {
```
Wait @deitch -- so what's the point of setting -f flag if we're not continuously copying the stream to stdout here?

deitch replied:
We are. `io.Copy` will keep going until it receives an `io.EOF` from its `Reader`, in this case `response.Body`. As long as it has more data to send - which it will do if the server never sends a "complete" message, it will keep doing it.

Personally, I like the fact that I don't have to worry about it on the client side at all, it just "does the right thing"

rvs replied:
Awesome! Sorry I missed that subtlety -- let me test it with EVE now.

rvs commented on cmd/log.go:
```diff
@@ -0,0 +1,50 @@
+		if err != nil {
+			log.Fatalf("error reading URL %s: %v", u, err)
+		}
+		if _, err := io.Copy(os.Stdout, response.Body); err != nil {
```
Same question as above

deitch replied:
Same answer. :-)

rvs commented:
Hey @deitch -- looks like `-f` may need a bit more finetuning. Here's what I get

```
$ adam admin info --server https://localhost:6000 --insecure  --uuid 75ea17e8-8b27-467c-b407-65b24db2b10c -f | jq .
2020/04/29 13:30:40 error reading URL https://localhost:6000/admin/device/75ea17e8-8b27-467c-b407-65b24db2b10c/info: Get https://localhost:6000/admin/device/75ea17e8-8b27-467c-b407-65b24db2b10c/info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

Now, admittedly, I'm running it against a remote Adam -- but I don't think that should make any difference (and besides without -f I get all the data I need). Same issue with logs btw.

From a quick glance it seems to me that you need to send at least an "empty" response to a client so that the client receives the headers (at least) and doesn't exit.

deitch commented:
Looks like it, although I couldn’t get it to do the same error when I tried. What were your steps to reproduce it @rvs ?

rvs commented:
I'm running EVE+Adam combo and talking to the Adam that is running inside a VM from the host. Take a look at this PR in EVE https://github.com/lf-edge/eve/pull/941 With this PR in place you can repro it by doing:

```
make HV=adam-kvm live
make ACCEL=y run
make run-proxy
adam admin info --server https://localhost:6000 --insecure  --uuid XXXX -f
```

giggsoff commented:
Hi, @deitch . I faced with the same problem as @rvs with docker on my host (without EVE).
As a workaround, I use curl:
```
curl -k -i -H "Accept: application/json" -H "X-Stream: true" https://127.0.0.1:3333/admin/device/d8a5a394-af8e-422d-aa4e-c6f581b14bc3/logs
```
and I found another two problems:
1. Stream is chunked not by "files", but by some buffer (maybe in curl):
I get stream like this: ```{"devID":"d8a5a394-af8e-422d-aa4e-c6f581b14bc3"```...```\"pid\":\"-\", \"parti``` (break in the end). And after few moments (with new log) I get the ending of first file ```tion\"}``` and another break in the end. I try to use ```http.Flusher Flush()``` but without success for now. Maybe it is connected with curl?
2. When I kill curl and start another one, I receive nothing. And I noticed, that when 10 log files received (after my stop and start curl action), It starts showing logs.

deitch commented:
I am trying to reproduce it, will let you know.

deitch commented:
> Stream is chunked not by "files", but by some buffer

My first thought would have been to flush, just like you did. We need to dig deeper. I think we will have to fix the basics, then do a separate update for that.

> When I kill curl and start another one, I receive nothing

That is strange. I wonder if it has to do with the channels. 

Will look. Once I actually can get Roman's live test case running...

deitch commented:
I just made a few changes:

* put the command under `adam device logs ... ` and `adam device info ... ` since these should be under the `device` command
* added flushing the headers right away, as well as several other flushes, to help with the timeout issues; that still isn't complete, but getting there
* added a close notifier, so it can capture when you break the connection client side and not get caught up

It needs some additional testing, so please go ahead @giggsoff

giggsoff commented:
@deitch, I found the problem:
https://github.com/lf-edge/adam/blob/9ee6ccc8ad82973a53e0b48fcf98f53fc3a81d59/cmd/admin.go#L63-L68
It is wrong for stream: after 10 seconds it returns.

deitch commented:
Ready for you @giggsoff . I removed that timeout, but wanted to get the correct ones for the rest of it, so that it still will time out if it doesn't get the headers or TLS connection, etc.

deitch commented:
Let's get this in, and then we can add improvements (like extra lines). I also want more logging options, etc. All in future PRs.

deitch commented:
And a time-series db embedded, if Roman's suggestions work. All in future.
